pkg/roachpb/metadata.go

```diff
@@ -782,11 +782,29 @@ func (h *GCHint) IsEmpty() bool {
 
 // Merge forwards latest delete timestamp of the receiver (lhs) with timestamp of
 // argument (rhs).
+// leftEmpty and rightEmpty arguments are set based on MVCCStats of ranges
+// being merged.
 // Returns true if receiver state was changed.
-func (h *GCHint) Merge(rhs *GCHint) bool {
-	if h.LatestRangeDeleteTimestamp.Less(rhs.LatestRangeDeleteTimestamp) {
-		// If right hand side of merge is higher, then use its timestamp and signal
-		// the change to caller.
+func (h *GCHint) Merge(rhs *GCHint, leftEmpty, rightEmpty bool) bool {
+	if !h.LatestRangeDeleteTimestamp.IsEmpty() {
+		// Only merge hints if it is set on both sides or other side is empty.
+		if !rhs.LatestRangeDeleteTimestamp.IsEmpty() || rightEmpty {
+			if h.LatestRangeDeleteTimestamp.Less(rhs.LatestRangeDeleteTimestamp) {
+				// If right hand side of merge is higher, then use its timestamp.
+				h.LatestRangeDeleteTimestamp = rhs.LatestRangeDeleteTimestamp
+				return true
+			}
+			return false
+		}
+
+		// If no hint on RHS and range is not empty then we need to reset our hint.
+		h.LatestRangeDeleteTimestamp = hlc.Timestamp{}
+		return true
```

Comment:
nit: this can be simplified, for example:

```go
	// If either side has data but no hint, merged range can't have a hint.
	if (rhs.LatestRangeDeleteTimestamp.IsEmpty() && !rightEmpty) ||
		(h.LatestRangeDeleteTimestamp.IsEmpty() && !leftEmpty) {
		updated := h.LatestRangeDeleteTimestamp.IsSet()
		h.LatestRangeDeleteTimestamp = hlc.Timestamp{}
		return updated
	}
	// Otherwise, use the newest hint.
	return h.ForwardLatestRangeDeleteTimestamp(rhs.LatestRangeDeleteTimestamp)
```

We should update the comment to describe this behavior as well.

Reply:
Whoops, wrong position here -- the above suggestion would replace the entire method.